internal/server/websocketServer.go

```diff
@@ -0,0 +1,191 @@
+	watchChan       chan dstore.WatchEvent
+	websocketServer *http.Server
+	upgrader        websocket.Upgrader
+	logger          *slog.Logger
+	shutdownChan    chan struct{}
+}
+
+func NewWebSocketServer(shardManager *shard.ShardManager, watchChan chan dstore.WatchEvent, logger *slog.Logger) *WebsocketServer {
+	mux := http.NewServeMux()
+	srv := &http.Server{
+		Addr:              fmt.Sprintf(":%d", config.WebsocketPort),
+		Handler:           mux,
+		ReadHeaderTimeout: 5 * time.Second,
+	}
+
+	upgrader := websocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool { return true },
+	}
+
+	websocketServer := &WebsocketServer{
+		shardManager:    shardManager,
+		querywatcher:    querywatcher.NewQueryManager(logger),
+		ioChan:          make(chan *ops.StoreResponse, 1000),
+		watchChan:       watchChan,
+		websocketServer: srv,
+		upgrader:        upgrader,
+		logger:          logger,
+		shutdownChan:    make(chan struct{}),
+	}
+
+	mux.HandleFunc("/", websocketServer.WebsocketHandler)
+	mux.HandleFunc("/health", func(w http.ResponseWriter, r *http.Request) {
+		_, err := w.Write([]byte("OK"))
+		if err != nil {
+			return
+		}
+	})
+	return websocketServer
+}
+
+func (s *WebsocketServer) Run(ctx context.Context) error {
+	var wg sync.WaitGroup
+	var err error
+
+	websocketCtx, cancelWebsocket := context.WithCancel(ctx)
+	defer cancelWebsocket()
+
+	s.shardManager.RegisterWorker("wsServer", s.ioChan)
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		select {
+		case <-ctx.Done():
+		case <-s.shutdownChan:
+			err = diceerrors.ErrAborted
+			s.logger.Debug("Shutting down Websocket Server")
+		}
+
+		shutdownErr := s.websocketServer.Shutdown(websocketCtx)
+		if shutdownErr != nil {
+			s.logger.Error("Websocket Server shutdown failed:", slog.Any("error", err))
+			return
+		}
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		s.logger.Info("Websocket Server running", slog.String("port", s.websocketServer.Addr[1:]))
+		err = s.websocketServer.ListenAndServe()
+	}()
+
+	wg.Wait()
+	return err
+}
+
+func (s *WebsocketServer) WebsocketHandler(w http.ResponseWriter, r *http.Request) {
+	// upgrade http connection to websocket
+	conn, err := s.upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		return
+	}
+
+	// closing handshake
+	defer func() {
+		_ = conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, "close 1000 (normal)"))
+		conn.Close()
+	}()
+
+	for {
+		// read incoming message
+		_, msg, err := conn.ReadMessage()
+		if err != nil {
+			sendTextMessage(conn, []byte("error: command reading failed"))
+			continue
+		}
+
+		// parse message to dice command
+		redisCmd, err := utils.ParseWebsocketMessage(msg)
+		if err == diceerrors.ErrEmptyCommand {
+			continue
+		} else if err != nil {
+			sendTextMessage(conn, []byte("error: parsing failed"))
+			continue
+		}
+
+		if redisCmd.Cmd == Abort {
+			close(s.shutdownChan)
+			break
+		}
+
+		if unimplementedCommandsWebsocket[redisCmd.Cmd] {
```

Comment:
Did you try QWATCH? What is this not unimplemented?
I think you can reuse a lot of what @lucifercr07 did as thats compatible with WS.